fold/fold.go

```diff
@@ -0,0 +1,895 @@
+		}
+
+		if ij2 != subsequence { // add energy
+			w, err := W(i2, j2, foldContext)
+			if err != nil {
+				return DefaultStructure, fmt.Errorf("Multibranch: subsequence (%d, %d, %d): %w", start, end, k, err)
+			}
+			e_sum += w.Energy
+		}
+
+	}
+
+	if unpaired < 0 {
+		return DefaultStructure, fmt.Errorf("Multibranch: subsequence (%d, %d, %d): unpaired < 0", start, end, k)
+	}
+
+	// penalty for unmatched bp and multi-branch
+	a, b, c, d := foldContext.Energies.Multibranch.A, foldContext.Energies.Multibranch.B, foldContext.Energies.Multibranch.C, foldContext.Energies.Multibranch.D
+	e_Multibranch := a + b*float64(len(branches)) + c*float64(unpaired)
+
+	if unpaired == 0 {
+		e_Multibranch = a + d
+	}
+
+	// energy of min-energy neighbors
+	e := e_Multibranch + e_sum
+
+	// pointer to next structures
+	if helix {
+		// branches.pop()
+		branches = branches[:len(branches)-1]
+	}
+
+	return NucleicAcidStructure{Energy: e, Description: fmt.Sprintf("BIFURCATION:%dn/%dh", unpaired, branches_count), Inner: branches}, nil
+}
+
+// InternalLoop calculates the free energy of an internal loop.
+//
+// The first and last bp of both left and right sequences
+// are not themselves parts of the loop, but are the terminal
+// bp on either side of it. They are needed for when there's
+// a single internal looping bp (where just the mismatching
+// free energies are used)
+// Note that both left and right sequences are in 5' to 3' direction
+// This is adapted from the "Internal Loops" section of SantaLucia/Hicks, 2004
+// Args:
+//
+//		start:  The index of the start of structure on left side
+//		i1: The index to the right of start
+//		end:  The index of the end of structure on right side
+//		j1: The index to the left of end
+//	 foldContext: The FoldingContext for this sequence
+//
+// Returns the free energy associated with the internal loop
+func InternalLoop(start, i1, end, j1 int, foldContext FoldContext) (float64, error) {
+	loop_left := i1 - start - 1
+	loop_right := end - j1 - 1
+	loop_len := loop_left + loop_right
+
+	if loop_left < 1 {
+		return 0, fmt.Errorf("internal loop: subsequence (%d, %d, %d, %d): missing left part of the loop", start, i1, end, j1)
+
+	}
+	if loop_right < 1 {
+		return 0, fmt.Errorf("internal loop: subsequence (%d, %d, %d, %d): missing right part of the loop", start, i1, end, j1)
+	}
+
+	// single bp mismatch, sum up the two single mismatch pairs
+	if loop_left == 1 && loop_right == 1 {
+		mm_left := Stack(start, i1, end, j1, foldContext)
+		mm_right := Stack(i1-1, i1, j1+1, j1, foldContext)
+		return mm_left + mm_right, nil
+	}
+	var d_h, d_s, dG float64
+	// apply a penalty based on loop size
+	if en, ok := foldContext.Energies.InternalLoops[loop_len]; ok {
+		d_h, d_s = en.EnthalpyH, en.EntropyS
+		dG = DeltaG(d_h, d_s, foldContext.T)
+	} else {
+		// it's too large an internal loop, extrapolate
+		en := foldContext.Energies.InternalLoops[30]
```

Comment:
why this magic number `30`? Should probably be internal package constant.

Reply:
It is the max length we have for precomputed energies, above that we use JacobsonStockmayer.
Moving to a const.